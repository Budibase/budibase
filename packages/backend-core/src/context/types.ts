import { IdentityContext } from "@budibase/types"

// keep this out of Budibase types, don't want to expose context info
export type ContextMap = {
  tenantId?: string
  appId?: string
  identity?: IdentityContext
  environmentVariables?: Record<string, string>
  isScim?: boolean
  automationId?: string
}
