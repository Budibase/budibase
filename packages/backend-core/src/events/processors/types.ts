import { Event, Identity, Group } from "@budibase/types"

export enum EventProcessorType {
  POSTHOG = "posthog",
  LOGGING = "logging",
}

export interface EventProcessor {
  processEvent(
    event: Event,
    identity: Identity,
    properties: any,
    timestamp?: string | number
  ): Promise<void>
  identify(identity: Identity, timestamp?: string | number): Promise<void>
  identifyGroup(group: Group, timestamp?: string | number): Promise<void>
  shutdown(): void
}
