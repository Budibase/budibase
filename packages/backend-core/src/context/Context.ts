import { AsyncLocalStorage } from "async_hooks"
import { ContextMap } from "./types"

export default class Context {
  static storage = new AsyncLocalStorage<ContextMap>()

  static run<T>(context: ContextMap, func: () => T) {
    return Context.storage.run(context, () => func())
  }

  static get(): ContextMap {
    return Context.storage.getStore() as ContextMap
  }
}
