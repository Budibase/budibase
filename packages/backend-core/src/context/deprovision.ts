import {
  getGlobalUserParams,
  getAllApps,
  doWithDB,
  StaticDatabases,
} from "../db"
import { doWithGlobalDB } from "../tenancy"
import { App, Tenants, User, Database } from "@budibase/types"

const TENANT_DOC = StaticDatabases.PLATFORM_INFO.docs.tenants
const PLATFORM_INFO_DB = StaticDatabases.PLATFORM_INFO.name

async function removeTenantFromInfoDB(tenantId: string) {
  try {
    await doWithDB(PLATFORM_INFO_DB, async (infoDb: Database) => {
      const tenants = (await infoDb.get(TENANT_DOC)) as Tenants
      tenants.tenantIds = tenants.tenantIds.filter(id => id !== tenantId)

      await infoDb.put(tenants)
    })
  } catch (err) {
    console.error(`Error removing tenant ${tenantId} from info db`, err)
    throw err
  }
}

export async function removeUserFromInfoDB(dbUser: User) {
  await doWithDB(PLATFORM_INFO_DB, async (infoDb: Database) => {
    const keys = [dbUser._id!, dbUser.email]
    const userDocs = await infoDb.allDocs({
      keys,
      include_docs: true,
    })
    const toDelete = userDocs.rows.map((row: any) => {
      return {
        ...row.doc,
        _deleted: true,
      }
    })
    await infoDb.bulkDocs(toDelete)
  })
}

async function removeUsersFromInfoDB(tenantId: string) {
  return doWithGlobalDB(tenantId, async (db: any) => {
    try {
      const allUsers = await db.allDocs(
        getGlobalUserParams(null, {
          include_docs: true,
        })
      )
      await doWithDB(PLATFORM_INFO_DB, async (infoDb: any) => {
        const allEmails = allUsers.rows.map((row: any) => row.doc.email)
        // get the id docs
        let keys = allUsers.rows.map((row: any) => row.id)
        // and the email docs
        keys = keys.concat(allEmails)
        // retrieve the docs and delete them
        const userDocs = await infoDb.allDocs({
          keys,
          include_docs: true,
        })
        const toDelete = userDocs.rows.map((row: any) => {
          return {
            ...row.doc,
            _deleted: true,
          }
        })
        await infoDb.bulkDocs(toDelete)
      })
    } catch (err) {
      console.error(`Error removing tenant ${tenantId} users from info db`, err)
      throw err
    }
  })
}

async function removeGlobalDB(tenantId: string) {
  return doWithGlobalDB(tenantId, async (db: Database) => {
    try {
      await db.destroy()
    } catch (err) {
      console.error(`Error removing tenant ${tenantId} users from info db`, err)
      throw err
    }
  })
}

async function removeTenantApps(tenantId: string) {
  try {
    const apps = (await getAllApps({ all: true })) as App[]
    const destroyPromises = apps.map(app =>
      doWithDB(app.appId, (db: Database) => db.destroy())
    )
    await Promise.allSettled(destroyPromises)
  } catch (err) {
    console.error(`Error removing tenant ${tenantId} apps`, err)
    throw err
  }
}

// can't live in tenancy package due to circular dependency on db/utils
export async function deleteTenant(tenantId: string) {
  await removeTenantFromInfoDB(tenantId)
  await removeUsersFromInfoDB(tenantId)
  await removeGlobalDB(tenantId)
  await removeTenantApps(tenantId)
}
