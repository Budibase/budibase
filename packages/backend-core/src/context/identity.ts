import {
  IdentityContext,
  IdentityType,
  User,
  isCloudAccount,
  Account,
  AccountUserContext,
  UserContext,
  Ctx,
} from "@budibase/types"
import * as context from "."

export function getIdentity(): IdentityContext | undefined {
  return context.getIdentity()
}

export function doInIdentityContext(identity: IdentityContext, task: any) {
  return context.doInIdentityContext(identity, task)
}

// used in server/worker
export function doInUserContext(user: User, ctx: Ctx, task: any) {
  const userContext: UserContext = {
    ...user,
    _id: user._id as string,
    type: IdentityType.USER,
    hostInfo: {
      ipAddress: ctx.request.ip,
      // filled in by koa-useragent package
      userAgent: ctx.userAgent._agent.source,
    },
  }
  return doInIdentityContext(userContext, task)
}

// used in account portal
export function doInAccountContext(account: Account, task: any) {
  const _id = getAccountUserId(account)
  const tenantId = account.tenantId
  const accountContext: AccountUserContext = {
    _id,
    type: IdentityType.USER,
    tenantId,
    account,
  }
  return doInIdentityContext(accountContext, task)
}

export function getAccountUserId(account: Account) {
  let userId: string
  if (isCloudAccount(account)) {
    userId = account.budibaseUserId
  } else {
    // use account id as user id for self-hosting
    userId = account.accountId
  }
  return userId
}
