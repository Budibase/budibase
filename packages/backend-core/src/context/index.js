const env = require("../environment")
const { Headers } = require("../../constants")
const { SEPARATOR, DocumentTypes } = require("../db/constants")
const cls = require("./FunctionContext")
const { getCouch } = require("../db")
const { getProdAppID, getDevelopmentAppID } = require("../db/conversions")
const { isEqual } = require("lodash")

// some test cases call functions directly, need to
// store an app ID to pretend there is a context
let TEST_APP_ID = null

const ContextKeys = {
  TENANT_ID: "tenantId",
  APP_ID: "appId",
  // whatever the request app DB was
  CURRENT_DB: "currentDb",
  // get the prod app DB from the request
  PROD_DB: "prodDb",
  // get the dev app DB from the request
  DEV_DB: "devDb",
  DB_OPTS: "dbOpts",
}

exports.DEFAULT_TENANT_ID = "default"

exports.isDefaultTenant = () => {
  return exports.getTenantId() === exports.DEFAULT_TENANT_ID
}

exports.isMultiTenant = () => {
  return env.MULTI_TENANCY
}

// used for automations, API endpoints should always be in context already
exports.doInTenant = (tenantId, task) => {
  return cls.run(() => {
    // set the tenant id
    cls.setOnContext(ContextKeys.TENANT_ID, tenantId)

    // invoke the task
    return task()
  })
}

/**
 * Given an app ID this will attempt to retrieve the tenant ID from it.
 * @return {null|string} The tenant ID found within the app ID.
 */
exports.getTenantIDFromAppID = appId => {
  if (!appId) {
    return null
  }
  const split = appId.split(SEPARATOR)
  const hasDev = split[1] === DocumentTypes.DEV
  if ((hasDev && split.length === 3) || (!hasDev && split.length === 2)) {
    return null
  }
  if (hasDev) {
    return split[2]
  } else {
    return split[1]
  }
}

const setAppTenantId = appId => {
  const appTenantId = this.getTenantIDFromAppID(appId) || this.DEFAULT_TENANT_ID
  this.updateTenantId(appTenantId)
}

exports.doInAppContext = (appId, task) => {
  if (!appId) {
    throw new Error("appId is required")
  }
  return cls.run(() => {
    // set the app tenant id
    setAppTenantId(appId)

    // set the app ID
    cls.setOnContext(ContextKeys.APP_ID, appId)

    // invoke the task
    return task()
  })
}

exports.updateTenantId = tenantId => {
  cls.setOnContext(ContextKeys.TENANT_ID, tenantId)
}

exports.updateAppId = appId => {
  try {
    cls.setOnContext(ContextKeys.APP_ID, appId)
    cls.setOnContext(ContextKeys.PROD_DB, null)
    cls.setOnContext(ContextKeys.DEV_DB, null)
    cls.setOnContext(ContextKeys.CURRENT_DB, null)
    cls.setOnContext(ContextKeys.DB_OPTS, null)
  } catch (err) {
    if (env.isTest()) {
      TEST_APP_ID = appId
    } else {
      throw err
    }
  }
}

exports.setTenantId = (
  ctx,
  opts = { allowQs: false, allowNoTenant: false }
) => {
  let tenantId
  // exit early if not multi-tenant
  if (!exports.isMultiTenant()) {
    cls.setOnContext(ContextKeys.TENANT_ID, this.DEFAULT_TENANT_ID)
    return
  }

  const allowQs = opts && opts.allowQs
  const allowNoTenant = opts && opts.allowNoTenant
  const header = ctx.request.headers[Headers.TENANT_ID]
  const user = ctx.user || {}
  if (allowQs) {
    const query = ctx.request.query || {}
    tenantId = query.tenantId
  }
  // override query string (if allowed) by user, or header
  // URL params cannot be used in a middleware, as they are
  // processed later in the chain
  tenantId = user.tenantId || header || tenantId

  // Set the tenantId from the subdomain
  if (!tenantId) {
    tenantId = ctx.subdomains && ctx.subdomains[0]
  }

  if (!tenantId && !allowNoTenant) {
    ctx.throw(403, "Tenant id not set")
  }
  // check tenant ID just incase no tenant was allowed
  if (tenantId) {
    cls.setOnContext(ContextKeys.TENANT_ID, tenantId)
  }
}

exports.isTenantIdSet = () => {
  const tenantId = cls.getFromContext(ContextKeys.TENANT_ID)
  return !!tenantId
}

exports.getTenantId = () => {
  if (!exports.isMultiTenant()) {
    return exports.DEFAULT_TENANT_ID
  }
  const tenantId = cls.getFromContext(ContextKeys.TENANT_ID)
  if (!tenantId) {
    throw new Error("Tenant id not found")
  }
  return tenantId
}

exports.getAppId = () => {
  const foundId = cls.getFromContext(ContextKeys.APP_ID)
  if (!foundId && env.isTest() && TEST_APP_ID) {
    return TEST_APP_ID
  } else {
    return foundId
  }
}

function getDB(key, opts) {
  const dbOptsKey = `${key}${ContextKeys.DB_OPTS}`
  let storedOpts = cls.getFromContext(dbOptsKey)
  let db = cls.getFromContext(key)
  if (db && isEqual(opts, storedOpts)) {
    return db
  }
  const appId = exports.getAppId()
  const CouchDB = getCouch()
  let toUseAppId
  switch (key) {
    case ContextKeys.CURRENT_DB:
      toUseAppId = appId
      break
    case ContextKeys.PROD_DB:
      toUseAppId = getProdAppID(appId)
      break
    case ContextKeys.DEV_DB:
      toUseAppId = getDevelopmentAppID(appId)
      break
  }
  db = new CouchDB(toUseAppId, opts)
  try {
    cls.setOnContext(key, db)
    if (opts) {
      cls.setOnContext(dbOptsKey, opts)
    }
  } catch (err) {
    if (!env.isTest()) {
      throw err
    }
  }
  return db
}

/**
 * Opens the app database based on whatever the request
 * contained, dev or prod.
 */
exports.getAppDB = opts => {
  return getDB(ContextKeys.CURRENT_DB, opts)
}

/**
 * This specifically gets the prod app ID, if the request
 * contained a development app ID, this will open the prod one.
 */
exports.getProdAppDB = opts => {
  return getDB(ContextKeys.PROD_DB, opts)
}

/**
 * This specifically gets the dev app ID, if the request
 * contained a prod app ID, this will open the dev one.
 */
exports.getDevAppDB = opts => {
  return getDB(ContextKeys.DEV_DB, opts)
}
