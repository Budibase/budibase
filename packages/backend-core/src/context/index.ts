export { DEFAULT_TENANT_ID } from "../constants"
export * as identity from "./identity"
export * from "./mainContext"
