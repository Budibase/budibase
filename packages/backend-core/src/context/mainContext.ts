// some test cases call functions directly, need to
// store an app ID to pretend there is a context
import env from "../environment"
import Context from "./Context"
import * as conversions from "../docIds/conversions"
import { getDB } from "../db/db"
import {
  DocumentType,
  SEPARATOR,
  StaticDatabases,
  DEFAULT_TENANT_ID,
} from "../constants"
import { Database, IdentityContext } from "@budibase/types"
import { ContextMap } from "./types"

let TEST_APP_ID: string | null = null

export function getGlobalDBName(tenantId?: string) {
  // tenant ID can be set externally, for example user API where
  // new tenants are being created, this may be the case
  if (!tenantId) {
    tenantId = getTenantId()
  }
  return baseGlobalDBName(tenantId)
}

export function getAuditLogDBName(tenantId?: string) {
  if (!tenantId) {
    tenantId = getTenantId()
  }
  if (tenantId === DEFAULT_TENANT_ID) {
    return StaticDatabases.AUDIT_LOGS.name
  } else {
    return `${tenantId}${SEPARATOR}${StaticDatabases.AUDIT_LOGS.name}`
  }
}

export function baseGlobalDBName(tenantId: string | undefined | null) {
  if (!tenantId || tenantId === DEFAULT_TENANT_ID) {
    return StaticDatabases.GLOBAL.name
  } else {
    return `${tenantId}${SEPARATOR}${StaticDatabases.GLOBAL.name}`
  }
}

export function getPlatformURL() {
  return env.PLATFORM_URL
}

export function isMultiTenant() {
  return !!env.MULTI_TENANCY
}

export function isTenantIdSet() {
  const context = Context.get()
  return !!context?.tenantId
}

export function isTenancyEnabled() {
  return env.MULTI_TENANCY
}

/**
 * Given an app ID this will attempt to retrieve the tenant ID from it.
 * @return The tenant ID found within the app ID.
 */
export function getTenantIDFromAppID(appId: string) {
  if (!appId) {
    return undefined
  }
  if (!isMultiTenant()) {
    return DEFAULT_TENANT_ID
  }
  const split = appId.split(SEPARATOR)
  const hasDev = split[1] === DocumentType.DEV
  if ((hasDev && split.length === 3) || (!hasDev && split.length === 2)) {
    return undefined
  }
  if (hasDev) {
    return split[2]
  } else {
    return split[1]
  }
}

function updateContext(updates: ContextMap): ContextMap {
  let context: ContextMap
  try {
    context = Context.get()
  } catch (err) {
    // no context, start empty
    context = {}
  }
  context = {
    ...context,
    ...updates,
  }
  return context
}

async function newContext(updates: ContextMap, task: any) {
  // see if there already is a context setup
  let context: ContextMap = updateContext(updates)
  return Context.run(context, task)
}

export async function doInAutomationContext(params: {
  appId: string
  automationId: string
  task: any
}): Promise<any> {
  const tenantId = getTenantIDFromAppID(params.appId)
  return newContext(
    {
      tenantId,
      appId: params.appId,
      automationId: params.automationId,
    },
    params.task
  )
}

export async function doInContext(appId: string, task: any): Promise<any> {
  const tenantId = getTenantIDFromAppID(appId)
  return newContext(
    {
      tenantId,
      appId,
    },
    task
  )
}

export async function doInTenant<T>(
  tenantId: string | null,
  task: () => T
): Promise<T> {
  // make sure default always selected in single tenancy
  if (!env.MULTI_TENANCY) {
    tenantId = tenantId || DEFAULT_TENANT_ID
  }

  const updates = tenantId ? { tenantId } : {}
  return newContext(updates, task)
}

export async function doInAppContext(
  appId: string | null,
  task: any
): Promise<any> {
  if (!appId && !env.isTest()) {
    throw new Error("appId is required")
  }

  let updates: ContextMap
  if (!appId) {
    updates = { appId: "" }
  } else {
    const tenantId = getTenantIDFromAppID(appId)
    updates = { appId }
    if (tenantId) {
      updates.tenantId = tenantId
    }
  }
  return newContext(updates, task)
}

export async function doInIdentityContext(
  identity: IdentityContext,
  task: any
): Promise<any> {
  if (!identity) {
    throw new Error("identity is required")
  }

  const context: ContextMap = {
    identity,
  }
  if (identity.tenantId) {
    context.tenantId = identity.tenantId
  }
  return newContext(context, task)
}

export function getIdentity(): IdentityContext | undefined {
  try {
    const context = Context.get()
    return context?.identity
  } catch (e) {
    // do nothing - identity is not in context
  }
}

export function getTenantId(): string {
  if (!isMultiTenant()) {
    return DEFAULT_TENANT_ID
  }
  const context = Context.get()
  const tenantId = context?.tenantId
  if (!tenantId) {
    throw new Error("Tenant id not found")
  }
  return tenantId
}

export function getAutomationId(): string | undefined {
  const context = Context.get()
  return context?.automationId
}

export function getAppId(): string | undefined {
  const context = Context.get()
  const foundId = context?.appId
  if (!foundId && env.isTest() && TEST_APP_ID) {
    return TEST_APP_ID
  } else {
    return foundId
  }
}

export const getProdAppId = () => {
  const appId = getAppId()
  if (!appId) {
    throw new Error("Could not get appId")
  }
  return conversions.getProdAppID(appId)
}

export function doInEnvironmentContext(
  values: Record<string, string>,
  task: any
) {
  if (!values) {
    throw new Error("Must supply environment variables.")
  }
  const updates = {
    environmentVariables: values,
  }
  return newContext(updates, task)
}

export function doInScimContext(task: any) {
  const updates: ContextMap = {
    isScim: true,
  }
  return newContext(updates, task)
}

export function getEnvironmentVariables() {
  const context = Context.get()
  if (!context.environmentVariables) {
    return null
  } else {
    return context.environmentVariables
  }
}

export function getGlobalDB(): Database {
  const context = Context.get()
  if (!context || (env.MULTI_TENANCY && !context.tenantId)) {
    throw new Error("Global DB not found")
  }
  return getDB(baseGlobalDBName(context?.tenantId))
}

export function getAuditLogsDB(): Database {
  if (!getTenantId()) {
    throw new Error("No tenant ID found - cannot open audit log DB")
  }
  return getDB(getAuditLogDBName())
}

/**
 * Gets the app database based on whatever the request
 * contained, dev or prod.
 */
export function getAppDB(opts?: any): Database {
  const appId = getAppId()
  if (!appId) {
    throw new Error("Unable to retrieve app DB - no app ID.")
  }
  return getDB(appId, opts)
}

/**
 * This specifically gets the prod app ID, if the request
 * contained a development app ID, this will get the prod one.
 */
export function getProdAppDB(opts?: any): Database {
  const appId = getAppId()
  if (!appId) {
    throw new Error("Unable to retrieve prod DB - no app ID.")
  }
  return getDB(conversions.getProdAppID(appId), opts)
}

/**
 * This specifically gets the dev app ID, if the request
 * contained a prod app ID, this will get the dev one.
 */
export function getDevAppDB(opts?: any): Database {
  const appId = getAppId()
  if (!appId) {
    throw new Error("Unable to retrieve dev DB - no app ID.")
  }
  return getDB(conversions.getDevelopmentAppID(appId), opts)
}

export function isScim(): boolean {
  const context = Context.get()
  const scimCall = context?.isScim
  return !!scimCall
}
