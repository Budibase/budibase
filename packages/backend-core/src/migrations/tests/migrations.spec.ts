import { testEnv, DBTestConfiguration } from "../../../tests"
import * as migrations from "../index"
import * as context from "../../context"
import { MigrationType } from "@budibase/types"

testEnv.multiTenant()

describe("migrations", () => {
  const config = new DBTestConfiguration()

  const migrationFunction = jest.fn()

  const MIGRATIONS = [
    {
      type: MigrationType.GLOBAL,
      name: "test" as any,
      fn: migrationFunction,
    },
  ]

  beforeEach(() => {
    config.newTenant()
  })

  afterEach(async () => {
    jest.clearAllMocks()
  })

  const migrate = () => {
    return migrations.runMigrations(MIGRATIONS, {
      tenantIds: [config.tenantId],
    })
  }

  it("should run a new migration", async () => {
    await config.doInTenant(async () => {
      await migrate()
      expect(migrationFunction).toHaveBeenCalled()
      const db = context.getGlobalDB()
      const doc = await migrations.getMigrationsDoc(db)
      expect(doc.test).toBeDefined()
    })
  })

  it("should match snapshot", async () => {
    await config.doInTenant(async () => {
      await migrate()
      const doc = await migrations.getMigrationsDoc(context.getGlobalDB())
      expect(doc).toMatchSnapshot()
    })
  })

  it("should skip a previously run migration", async () => {
    await config.doInTenant(async () => {
      const db = context.getGlobalDB()
      await migrate()
      const previousDoc = await migrations.getMigrationsDoc(db)
      await migrate()
      const currentDoc = await migrations.getMigrationsDoc(db)
      expect(migrationFunction).toHaveBeenCalledTimes(1)
      expect(currentDoc.test).toBe(previousDoc.test)
    })
  })
})
