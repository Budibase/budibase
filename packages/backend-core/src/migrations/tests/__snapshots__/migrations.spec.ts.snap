// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrations should match snapshot 1`] = `
{
  "_id": "migrations",
  "_rev": "1-2f64479842a0513aa8b97f356b0b9127",
  "createdAt": "2020-01-01T00:00:00.000Z",
  "test": 1577836800000,
  "updatedAt": "2020-01-01T00:00:00.000Z",
}
`;
